### PR TITLE
Overload Forward() function.

### PR DESCRIPTION
**Pull request by akhandait**
This allows users to forward pass partially through the network.

### CODE REVIEW

File: src/mlpack/methods/ann/ffn.hpp [hunk 1]
```diff
@@ -284,6 +284,19 @@ class FFN
...
+   *
+   * This function is meant for the cases when users require a forward pass only
+   * through certain layers and not the entire network.
+   *
+   * @param inputs The input data for the specified first layer.
+   * @param results The predicted results from the specified last layer.
+   * @param begin The index of the first layer.
+   * @param end The index of the last layer.
+   */
+  void Forward(arma::mat inputs, arma::mat& results, size_t begin, size_t end);
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: Can we use ``const size_t``, to be consistent with the rest of the codebase?

**Comment by zoq** [comment]
Ah, good point, yeah that would be tremendously helpful.

**Comment by akhandait** [comment]
@rcurtin I think that's a good idea. I will add a test in `FeedForwardNetworkTest`, or should I add it in `ANNLayerTest`?

**Comment by rcurtin** [comment]
No preference from my end, I think anywhere is fine. Thanks! :)

**Comment by akhandait** [comment]
@rcurtin I understand, I will add a better test for this soon. I see that both of you have approved, but I think we should merge this after I add a better test.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 2]
```diff
@@ -620,4 +620,47 @@ BOOST_AUTO_TEST_CASE(CustomLayerTest)
...
+
+  // As we only forward pass through Add module, input and output should
+  // differ by a matrix of ones.
+  CheckMatrices(input, output - 1);
+
+  // Forward pass only through the Add module and the Identity layer.
+  model.Forward(input,
+                output,
+                1 /* Index of the Add module */,
+                2 /* Index of the Identity layer */);
```
Comment (zoq) [inline src/mlpack/tests/feedforward_network_test.cpp]: Do you mind to add another test, that includes the linear layer as well into the output, we can set the weights to ones as well; what I'd like to see is that the output changed. Let's suppose the end index doesn't work as expected the output is still the same as the first one since the second layer is an Identity layer.

**Comment by akhandait** [comment]
@rcurtin What is the comment on memory usage?

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 3]
```diff
@@ -620,4 +620,51 @@ BOOST_AUTO_TEST_CASE(CustomLayerTest)
...
+{
+  FFN<NegativeLogLikelihood<>, RandomInitialization> model;
+  model.Add<Linear<> >(5, 10);
+
+  // Add a new Add<> module which adds a constant term to the input.
+  Add<>* AddModule = new Add<>(10);
+  model.Add(AddModule);
+
+  LinearNoBias<>* LinearNoBiasModule = new LinearNoBias<>(10, 10);
+  model.Add(LinearNoBiasModule);
```
Comment (rcurtin) [inline src/mlpack/tests/feedforward_network_test.cpp]: Weird, the comment didn't seem to post.  I wonder what I did wrong.

Anyway: whenever you use `new`, you have to also use `delete` otherwise there is a memory leak.  So at the end of this test you'd need to add that.  But I would say, in this case, you can actually do something a little easier and avoid calling `new` at all:

```
Add<> addModule(10);
model.Add(addModule);
```

(Also note the lowercase variables, please do that to match the style guide. :))

Sorry this didn't post earlier!

**Comment by akhandait** [comment]
@rcurtin I don't it's possible here to do this
```cpp
Add<> addModule(10);
model.Add(addModule);
```
Add() expects a pointer and throws a error if we try the above.

**Comment by zoq** [comment]
Did you test ``model.Add<Add<> >(addModule);`` as well?

**Comment by rcurtin** [comment]
I opened #1489 to discuss that issue, so I think it's fine to leave it as a layer created with `new` for now.

File: src/mlpack/tests/feedforward_network_test.cpp [hunk 4]
```diff
@@ -620,4 +620,55 @@ BOOST_AUTO_TEST_CASE(CustomLayerTest)
   arma::mat prediction = arma::zeros<arma::mat>(1, predictionTemp.n_cols);
 }
 
+/**
+ * Test the overload of Forward function which allows partial forward pass.
+ */
+BOOST_AUTO_TEST_CASE(PartialForwardTest)
+{
+  FFN<NegativeLogLikelihood<>, RandomInitialization>* model =
+      new FFN<NegativeLogLikelihood<>, RandomInitialization>();
```
Comment (rcurtin) [inline src/mlpack/tests/feedforward_network_test.cpp]: Ok, now I understand what's happened here.  Changing `model` to be a manually-allocated pointer to an `FFN` isn't necessary, my concern originally was that I saw that `addModule` and `linearNoBiasModule` were being allocated without any `delete` later.  So I thought that the memory that was being allocated would never be freed.

However after digging into it, the particular way this code currently works (see #1489), `model` will actually delete `addModule` and `linearNoBiasModule` when its destructor is called.  That destructor gets called regardless of whether `model` is allocated with `new` or not---when using `new`, the destructor is called when we call `delete`.  But if we use a local variable for `model`, then the destructor is called when it goes out of scope (at the end of the function) and the memory is freed there too.

So you can revert to a local `FFN` for `model` if you like, but this code works either way and doesn't leak right now.  Depending on what we do for #1489, this may need to be adjusted later, but I don't think we need to worry about it.

tl;dr: I was incorrect about the memory leak, sorry about that. :)

Comment (akhandait) [reply src/mlpack/tests/feedforward_network_test.cpp]: Okay, I guess it's okay for me to revert to the second last commit then. Yeah, even I forgot that the FFN class deletes everything.

**Comment by zoq** [comment]
Let me restart the travis build, just to see the test pass.

**Comment by zoq** [comment]
Thanks again for the great work!
